/*
 * Copyright (c) 2013 L2jMobius
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
 * IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package handlers.admincommandhandlers;

import org.l2jmobius.gameserver.handler.IAdminCommandHandler;
import org.l2jmobius.gameserver.model.WorldObject;
import org.l2jmobius.gameserver.model.actor.Creature;
import org.l2jmobius.gameserver.model.actor.Player;

/**
 * @author Skache
 */
public class AdminInvul implements IAdminCommandHandler
{
	private static final String[] ADMIN_COMMANDS =
	{
		"admin_invul",
		"admin_setinvul",
		"admin_undying",
		"admin_setundying"
	};
	
	@Override
	public boolean onCommand(String command, Player activeChar)
	{
		switch (command)
		{
			case "admin_invul":
			{
				toggleInvul(activeChar, activeChar);
				break;
			}
			case "admin_setinvul":
			{
				final WorldObject target = activeChar.getTarget();
				if ((target != null) && target.isPlayer())
				{
					toggleInvul(target.asPlayer(), activeChar);
				}
				else
				{
					activeChar.sendSysMessage("You must target a player.");
				}
				break;
			}
			case "admin_undying":
			{
				toggleUndying(activeChar, activeChar);
				AdminHtml.showAdminHtml(activeChar, "gm_menu.htm");
				break;
			}
			case "admin_setundying":
			{
				final WorldObject target = activeChar.getTarget();
				if ((target != null) && target.isCreature())
				{
					toggleUndying(target.asCreature(), activeChar);
				}
				else
				{
					activeChar.sendSysMessage("You must target a creature.");
				}
				break;
			}
		}
		
		return true;
	}
	
	private void toggleInvul(Player target, Player admin)
	{
		final boolean invul = !target.isInvul();
		target.setInvul(invul);
		
		// Notify target.
		target.sendSysMessage("You are now " + (invul ? "invulnerable." : "mortal."));
		
		// Notify admin.
		if (admin != target)
		{
			admin.sendSysMessage("You made " + target.getName() + " " + (invul ? "invulnerable." : "mortal."));
		}
	}
	
	private void toggleUndying(Creature target, Player admin)
	{
		final boolean undying = !target.isUndying();
		target.setUndying(undying);
		
		// Notify target if it's a player.
		if (target.isPlayer())
		{
			target.asPlayer().sendSysMessage("You are now " + (undying ? "undying." : "mortal."));
		}
		
		// Notify admin.
		if (admin != target)
		{
			admin.sendSysMessage("You made " + target.getName() + " " + (undying ? "undying." : "mortal."));
		}
	}
	
	@Override
	public String[] getCommandList()
	{
		return ADMIN_COMMANDS;
	}
}
